src/Engine/Registry.h

```diff
@@ -0,0 +1,208 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with OpenXcom.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <concepts>
+#include <entt/entt.hpp>
+#include <ranges>
+#include "..\Savegame\Country.h"
+#include "..\Mod\RuleCountry.h"
+
+namespace OpenXcom
+{
+
+class Registry
+{
+private:
+	entt::registry _registry{};
+public:
+	// Gets access to the underlying registry.
+	[[nodiscard]] entt::registry& raw() { return _registry; }
+	// Gets access to the underlying registry.
+	[[nodiscard]] const entt::registry& raw() const { return _registry; }
+
+	/**
+	 * @brief Lists all Components in the registry in a const context
+	 * @tparam Component the type of componet to list.
+	 * @return an itterable of const Component references in the registry
+	 */
+	template<typename Component>
+	[[nodiscard]] auto list() const
+	{
+		return _registry.view<const Component>() | std::views::elements<1>();
+	}
+
+	/**
+	 * @brief Lists all Components in the registry in a mutable context
+	 * @tparam Component the type of componet to list.
+	 * @return an itterable of mutable Component references in the registry
+	 */
+	template<typename Component>
+	[[nodiscard]] auto list()
+	{
+		return _registry.view<Component>().each() | std::views::elements<1>();
+	}
+
+	/**
+	 * @brief Lists with a filter all Components in the registry in a const context
+	 * @tparam Component the type of componet to list and filter.
+	 * @param predicate the const Component predicate to use to filter the list
+	 * @return an itterable of const Component references in the registry that pass the predicate
+	 */
+	template<typename Component>
+	[[nodiscard]] auto list(std::predicate<const Component> auto predicate) const
+	{
+		return list<Component>() | std::views::filter(predicate);
+	}
+
+	/**
+	 * @brief Lists with a filter all Components in the registry in a mutable context
+	 * @tparam Component the type of componet to list and filter.
+	 * @param predicate the const Component predicate to use to filter the list
+	 * @return an itterable of mutable Component references in the registry that pass the predicate
+	 */
+	template<typename Component>
+	[[nodiscard]] auto list(std::predicate<const Component> auto predicate)
+	{
+		return list<Component>() | std::views::filter(predicate);
+	}
+
+	/**
+	 * @brief Creates a new entity and emplaces a new Component to it
+	 * @tparam Component the type of Component to emplace
+	 * @tparam ...ConstructorArgs the type of arguments to pass to the Component constructor
+	 * @param ...args the arguments to be forward to the Component constructor
+	 * @return A reference to the newly created Component
+	*/
+	template<typename Component, typename... ConstructorArgs>
+	Component& createAndEmplace(ConstructorArgs&&... args)
```

Comment:
Will probably want to phase this out and create factories that do this sort of stuff for us eventually

Reply:
:+1: